policy/modules/services/kubernetes.if

```diff
@@ -0,0 +1,895 @@
+	allow $2 kubernetes_home_t:lnk_file { manage_lnk_file_perms relabel_lnk_file_perms };
+	userdom_user_home_dir_filetrans($2, kubernetes_home_t, dir, ".kube")
+
+	allow $3 $1_kubectl_t:process { ptrace signal_perms };
+	ps_process_pattern($3, $1_kubectl_t)
+
+	auth_use_nsswitch($1_kubectl_t)
+
+	# kubectl executes an editor when editing files.
+	# transition back to the user domain when running them
+	corecmd_bin_domtrans($1_kubectl_t, $2)
+
+	optional_policy(`
+		systemd_user_app_status($1, $1_kubectl_t)
+	')
+')
+
+#######################################
+## <summary>
+##	Execute kubelet in the kubelet domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`kubernetes_domtrans_kubelet',`
+	gen_require(`
+		type kubelet_t, kubelet_exec_t;
+	')
+
+	corecmd_search_bin($1)
+	domtrans_pattern($1, kubelet_exec_t, kubelet_t)
+')
+
+########################################
+## <summary>
+##	Execute kubelet in the kubelet domain,
+##	and allow the specified role the
+##	kubelet domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+## <param name="role">
+##	<summary>
+##	The role to be allowed the kubelet domain.
+##	</summary>
+## </param>
+#
+interface(`kubernetes_run_kubelet',`
+	gen_require(`
+		type kubelet_t;
+	')
+
+	role $2 types kubelet_t;
+
+	kubernetes_domtrans_kubelet($1)
+')
+
+########################################
+## <summary>
+##	Connect to kubelet over a unix stream socket.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`kubernetes_stream_connect_kubelet',`
+	gen_require(`
+		type kubelet_t;
+		type kubernetes_runtime_t;
+	')
+
+	files_search_runtime($1)
+	stream_connect_pattern($1, kubernetes_runtime_t, kubernetes_runtime_t, kubelet_t)
+	allow $1 kubernetes_runtime_t:sock_file read_sock_file_perms;
+')
+
+#######################################
+## <summary>
+##	Read the process state (/proc/pid)
+##	of kubelet.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`kubernetes_read_kubelet_state',`
+	gen_require(`
+		type kubelet_t;
+	')
+
+	ps_process_pattern($1, kubelet_t)
+')
+
+#######################################
+## <summary>
+##	Inherit and use file descriptors from
+##	kubelet.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`kubernetes_use_kubelet_fds',`
+	gen_require(`
+		type kubelet_t;
+	')
+
+	allow $1 kubelet_t:fd use;
+')
+
+#######################################
+## <summary>
+##	Allow kubelet to send a kill signal
+##	to the specified domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`kubernetes_kubelet_sigkill',`
```

Comment:
I've made this change, but I think what we actually want is something like `kubernetes_kubelet_kill`? This interface is to allow `kubelet_t` to kill another domain, not the other way around.

Reply:
> This interface is to allow `kubelet_t` to kill another domain, not the other way around.

Hmm, I missed this detail. This can't be included somewhere else, e.g. in one of the attribute uses?

Reply:
We can add this access to `kubernetes_container_engine_domain`. I don't know if any other container engines need this access, but I also don't see the harm in it.

Reply:
Actually, `crio_conmon_t` is not an engine domain, so there isn't an attribute we can use for this. I assume what actually happens is that when the kubelet wants to forcibly kill a container, it will send a kill signal to every process in the container's cgroup.

I don't really see another clean way to add this access.

Reply:
I've changed the name of this interface to `kubernetes_kubelet_kill()`.

Another alternative is create a new attribute for this access, i.e. `kubernetes_kubelet_managed_domain`. This would also allow us to condense the many calls to `ps_process_pattern()` that were added.

Reply:
>This would also allow us to condense the many calls to ps_process_pattern() that were added.

Scratch that. This would actually give too much access as not all of the domains that kubelet needs to read the state of need to be killed by it.